### PR TITLE
[geometry] Deprecate geometry data as State in SceneGraph

### PR DESCRIPTION
**Pull request by SeanCurtis-TRI**
This changes the default behavior in SceneGraph; a default-constructed SceneGraph stores its geometry data as a Parameter (previously it was State).

This leaves a stop-gap constructor in place (marked as deprecated) to explicitly request the data to be stored as State.

Related to #9501 and #14189.

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14245)

### CODE REVIEW

**Review by SeanCurtis-TRI — COMMENTED** [review]
Testing to see if changing State to Parameter has any impact in Drake or Anzu (see Anzu build: Anzu-master-Drake-PR 84). I anticipate it should be perfectly happy. If so, we'll proceed with review.

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14245#-:-MKL2VGb2RMmnGj00Yp9:b-xblxfz)* status: needs platform reviewer assigned, needs at least two assigned reviewers, labeled "do not merge"

**Review by SeanCurtis-TRI — COMMENTED** [review]
Hmmm....two different runs against Anzu produced two different sets of timeouts -- on disjoint sets. I'm disinclined to believe that has anything to do with this PR, but I'm going to assign reviewers that will have a better basis for arguing otherwise than my rather naive skepticism.

+@jwnimmer-tri for feature review (as of Monday, please).

-(status: do not merge) -(status: do not review) 

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14245#-:-MKLTexE2t_d9omd7QWK:b-ueadbv)* status: LGTM missing from assignee jwnimmer-tri(platform), needs at least two assigned reviewers, labeled "do not merge" (waiting on @jwnimmer-tri)

**Review by jwnimmer-tri — COMMENTED** [review]
:lgtm: feature.

> Reviewed 5 of 5 files at r1.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14245#-:-MKQNaKm6U8NfgkDDEjM:b-n37teh)* status: needs at least two assigned reviewers

**Review by SeanCurtis-TRI — COMMENTED** [review]
+@ericcousineau-tri for platform review, please.

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14245#-:-MK_47oJBMPCubAvOsx7:bvtjo57)* status: LGTM missing from assignee EricCousineau-TRI(platform) (waiting on @EricCousineau-TRI)

**Review by EricCousineau-TRI — COMMENTED** [review]
> Reviewed 1 of 5 files at r1.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14245)* status: 1 unresolved discussion, LGTM missing from assignee EricCousineau-TRI(platform) (waiting on @SeanCurtis-TRI)

*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake/14245#-MKazjtf5PKEm5YPupDo:-MKazjthCTR2-5SHffyE:b-62b6ev) (no related file):*
nit My question as always: Is there an issue for this, or can the motivation be summarized in the commit?

From the diff, this is my starting stub:
"Before, geometry data could be modified via an unrestricted update, and may <confuse users because?>"

I could understand updating default state in an unrestricted update being a bad thing, but... we can do the same thing with other things, like `MultibodyPlant::SetDefaultPose`?

I'm happy to just have an issue to look at.

---

**Review by EricCousineau-TRI — COMMENTED** [review]
:lgtm:

> Reviewed 4 of 5 files at r1.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14245#-:-MKb-nQw2Bg0YzOhMOB9:bnfp4nl)* status: 1 unresolved discussion (waiting on @SeanCurtis-TRI)

*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake/14245#-MKazjtf5PKEm5YPupDo:-MKb-oXs3BgcWYOu7iF7:b-jzobj5) (no related file):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

nit My question as always: Is there an issue for this, or can the motivation be summarized in the commit?

From the diff, this is my starting stub:
"Before, geometry data could be modified via an unrestricted update, and may <confuse users because?>"

I could understand updating default state in an unrestricted update being a bad thing, but... we can do the same thing with other things, like `MultibodyPlant::SetDefaultPose`?

I'm happy to just have an issue to look at.
</blockquote></details>

(Feel free to merge and address this afterwards and/or never ;)

---

**Comment by jwnimmer-tri** [comment]
The PR that introduced the non-deprecated constructor linked to the issue.  That was PR #14189 for issue #9501.

**Review by SeanCurtis-TRI — COMMENTED** [review]
Thanks!

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14245#-:-MKb56xtA0b1h6ah3LRW:b-ucfknh)* status: :shipit: complete! all discussions resolved, LGTM from assignees jwnimmer-tri(platform),EricCousineau-TRI(platform)

*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake/14245#-MKazjtf5PKEm5YPupDo:-MKb4aFV5xV4Zzk0eCtB:bdxwg3b) (no related file):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

(Feel free to merge and address this afterwards and/or never ;)
</blockquote></details>

I'll update the commit message as I merge. Short version is that geometry data as `SceneGraph` made an otherwise continuous system hybrid (abstract state). This messes with things. The only real value to having it as `State` was to be able to modify via event (which made sense when we thought we were going to be doing cars), but not with robots. Particularly given the fact that even if it *made* sense, we still don't have a meaningful mechanism for one system to trigger an event in another system.

So, tl;dr:

Geometry data state caused problems without value. To simplify the world, we're going to keep it as a Parameter until we have a proven need to move it.

---

**Comment by EricCousineau-TRI** [comment]
*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake/14245#-MKazjtf5PKEm5YPupDo:-MKb5550DdCJpQYFsuJL:b-sa9cos) (no related file):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

I'll update the commit message as I merge. Short version is that geometry data as `SceneGraph` made an otherwise continuous system hybrid (abstract state). This messes with things. The only real value to having it as `State` was to be able to modify via event (which made sense when we thought we were going to be doing cars), but not with robots. Particularly given the fact that even if it *made* sense, we still don't have a meaningful mechanism for one system to trigger an event in another system.

So, tl;dr:

Geometry data state caused problems without value. To simplify the world, we're going to keep it as a Parameter until we have a proven need to move it.
</blockquote></details>

OK Thanks y'all!

I see this comment in particular:
https://github.com/RobotLocomotion/drake/issues/9501#issuecomment-606064329

My understanding:
Basically, keep `State` clean,  demote geometry data to a `Parameter`; that now places the burden of "AbstractValue pollution" on parameter-based optimization, but TTBOMK that's a less cumbersome issue than it polluting state.

---
